File: 3/main.java
// sliding window optimized (hash table)

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length(), result = 0;
        if(s == null) {
            return result;
        }
        Map<Character, Integer> map = new HashMap<>();
        
        for(int j = 0, i = 0; j < n; j++) {
            if(map.containsKey(s.charAt(j))) {
                i = Math.max(i, map.get(s.charAt(j)));
            }
            result = Math.max(result, j - i + 1);
            map.put(s.charAt(j), j + 1);
        }
        
        return result;
    }
}

File: 3/main.py
# sliding window optimized (hash table)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n, result = len(s), 0
        if s is None:
            return result
        hashMap = {}
        i = 0
        
        for j in range(n):
            if s[j] in hashMap:
                i = max(i, hashMap[s[j]])
            result = max(result, j - i + 1)
            hashMap[s[j]] = j + 1
            
        return result
